LiS/man/man4/connld.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id$
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date$ by $Author$
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: connld.4.man,v $
.\" Revision 1.1.4.4  2008-04-29 08:33:18  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH CONNLD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B connld
\- \fISTREAMS\fR module supporting the generation of unique new \fISTREAMS\fR \fIpipes\fR from an existing pipe end
.\"
.\"
.SH SYNOPSIS
.PP
.HP
.BI "int " pfds[2] ;
.HP
.BI "int s_pipe(" pfds );
.HP
.BI "int ioctl(" pfds[0] ", " I_PUSH ", " \*(lqconnld\*(rq );
.HP
.BI "int fattach(" pfds[0] ", " \*(lq/tmp/server\*(rq );
.HP
.BI "int ioctl(" pfds[1] ", " I_RECVFD ", struct strfdinser *" fdi );
.\"
.\"
.SH DESCRIPTION
.PP
The
.B connld
module provides a means to generate multiple unique
.I STREAMS
.I pipes
from a single existing
.I pipe
end.
.sb
.B connld
may only be
.I pushed
(via the
.I STREAMS
.B I_PUSH
.IR ioctl )
onto a
.I STREAMS
.IR pipe .
When first
.IR pushed ,
.B connld
does nothing; on each subsequent
.BR open "(2), "
.B connld
will generate a unique
.I STREAMS
.IR pipe .
One end of each new pipe replaces the original pipe end from the
perspective of the
.B open
call.
The other end of each new pipe is sent, effectively as if by
the
.B I_SENDFD
.IR ioctl ,
to the other end of the original pipe, ostensibly to be received
by a subsequent
.B I_RECVFD
.I ioctl
operation.
.\"
.\"
.SH "APPLICATION USAGE"
.PP
The intent of
.B connld
is to provide a means to generate unique pipes which separately
and independently connect client processes to a server process.
The point of access for such clients is expected to be a path
name known to all such clients and to which a pipe end may be
connected (via
.BR fattach "(3)) "
by the server process.
The server establishes the original pipe, pushes
.B connld
onto the client end, and then listens via
.B I_RECVFD
for new connections on the server end.
A client wishing to connect to the server will
.BR open "(2) "
the path name representing the client end, and can determine via
.BR isastream "(3) "
whether or not the server process is active and attached.  If 
it is, the
.BR open "() "
call returns one end of a unique new pipe that thus connects
the client to the server.
.sp
Such a server is responsible both for accepting new connections
via
.B I_RECVFD
on the original pipe, and for communicating with clients so
connected via the received pipe ends.
It would also be reasonable for such a server process to invalidate
the point of access by calling
.BR fdetach "(3) "
before terminating.
.sp
It should be noted that the
.BR poll "(2) "
primitive may be used to indicate when an 
.B M_PASSFP
representing a newly passed file is available on the original server
pipe end.
This is reflected by the
.B POLLIN
status setting in the
.B events
and
.B revents
fields of a
.B pollfd
structure.
Moreover, any attempt to read an
.B M_PASSFP
message via the data-receiving primitives (i.e., 
.BR read "(2), " getmsg "(2), and " getpmsg "(2s)) "
will fail with
.BR errno "(3) "
returning an
.RB [ EBADMSG ]
indication without discarding the message.
.sp
Even so, it should be reasonable to expect only
.B M_PASSFP
messages will be received on the original server pipe end,
since it is not possible to carry on normal data traffic which
has
.B connld
on one end, since
.B connld
does not support such traffic.
.sp
The use of
.B connld
can be made entirely free-standing by attaching well-known paths to
.I both
ends of the original pipe.  The relevant capabilities are implemented
in
.B LiS
.[
LiS
.]
so that the original creator of the
.B pipe
can close both ends after attaching paths to them, and the
process of passing file descriptors can still be carried out via
new
.BR open "()'s "
as long as both ends remain attached.
.\"
.\"
.SH MODULES
.PP
.B connld
is a
.I STREAMS
module named
.RI \*(lq connld \*(rq.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR fattach "(3), "
.BR fattach "(8), "
.BR fdetach "(3), "
.BR fifo "(4), "
.BR pipe "(3), "
.BR STREAMS "(4) "
.\"
.\"
.SH HISTORY
.PP
UNIX\(rg\ System V Release 4 (SVR4)
.[
svr4
.]
.\"
.\"
.SH AUTHOR
.PP
John Boyd, protologos LLC.
.[
$LIST$
.]
.T3
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77


NEWS
OpenSS7 Master Package -- history of user visible changes.  2009-05-04
$Id: NEWS,v 0.9.2.72 2009-04-19 00:28:25 brian Exp $
Copyright (c) 2008-2009  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
`OpenSS7 Master Package' package.

 Major changes for release openss7-0.9.2.F.20090405
---------------------------------------------------

This is the seventh `OpenSS7 Project' release of the `OpenSS7 Master
Package' package.  For a sub-package listing, see `README'
(http://www.openss7.org/codefiles/openss7/README) in the release.  Most
of the sub-packages in the release are production grade for `Linux
Fast-STREAMS'.(1)  All existing validation test suites run clean on
supported distributions, production kernels and architectures.

The original intention was that the `OpenSS7 Master Package' package
not be released as often as subpackages and, indeed, newer subpackages
can be strapped into the master package and built automatically.
However, in reality, the `OpenSS7 Master Package' package has been
released as often as the subpackages.  To rebuild the master package
with a new sub-package release, simply replace the directory to which
the sub-package belongs with the unpackged sub-package reelase and then
rebuild the master package.  This release provides support for recent
distributions, production kernels and tool chains.  The release also
provides support for significiant new features in the sub-packages.

Major features since the last public release are:

   - Deprecated sub-packages have been excluded from the master package
     release.  The excluded sub-packages are:

         - `sctp-0.2.27.20090405'
         - `iperf-2.0.8.20090405'
         - `netperf-2.3.7.20090405'
         - `osr61-0.9.2.3.20090405'

   - Documentation license upgraded to GFDL 1.3.

   - Updated tool chain to `m4-1.4.13', `automake-1.10.2',
     `bison-2.4.1', `libtool-2.2.6a', `texinfo-4.13a'.

   - Addition of the following sub-packages:

         - `stratm-0.9.2.1.20090405'

   - Addition of the `stratm' sub-package to the master build and
     release that now includes all of the ATM components that were
     previously part of the `strss7' sub-package.

   - Significant feature updates were made to:

         - `strchan-0.9.2.4.20090405'
         - `strsock-0.9.2.4.20090405'
         - `strnsl-0.9.2.4.20090405'
         - `sigtran-0.9.2.4.20090405'

   - Significant rework of the `strchan' sub-package into the `strss7',
     `strisdn', `strx25' and `striso' packages.  One single driver now
     meets the needs of all sub-packages.

   - Completion of the `strsock' package, moves from pre-alpha to beta.

   - The `strnsl' subpackage moves from alpha to production.

   - Significant completion of the `sigtran' package: all components
     moved back to production.


_This is a public, stable, production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
`rpm(8)' and `dpkg(1)' based distributions, and can be used on
production kernels without patching or recompiling the kernel.

This package is publicly released strictly under the `GNU Affero
General Public License Version 3'; however, commercial licensing is
available.  The release is available as an `autoconf(1)' tarball, SRPM,
DSC, and a set of binary RPMs and DEBs for popular distributions.  See
the package page (http://www.openss.org/openss7_pkg.html) for the
`autoconf(1)' tarballs, SRPMs, DSCs and binary RPMs and DEBs.  See the
repository page (http://www.openss7.org/repos.html) for information
concerning network installation and update sources.

See
`http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/ChangeLog'
and `http://www.openss7.org/codefiles/openss7-0.9.2.F.20090405/NEWS' in
the release for more information.  Also, see the `openss7.pdf' manual
in the release (also in html
(http://www.openss7.org/openss7_manual.html)).

For the news release, see `http://www.openss7.org/rel2009XXXX_L.html'.

---------- Footnotes ----------

(1) Sub-packages are no longer supported on the deprecated `LiS' STREAMS
package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - License upgrade to `AGPL Version 3'.

   - Modifications to build under `Fedora' `2.6.22.5-49' kernel.  These
     changes also support `2.6.22.9-91.fc7' kernel.  Modifications to
     build under `Fedora' `2.6.25-45.fc9' and `2.6.26.5-45.fc9' kernels.
     Documented `lib32gcc1' problem on Ubuntu.  Noted problem running
     under XEN kernels.  XEN kernels are not yet supported.  Added
     `MODULE_VERSION' to all modules and drivers.

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.  Improvements to common build
     process for embedded and cross-compile targets.  Cross-compile

streams/src/modules/connld.c
/*****************************************************************************

 @(#) $RCSfile: connld.c,v $ $Name:  $($Revision: 0.9.2.43 $) $Date: 2008-09-22 20:31:31 $

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2009  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-09-22 20:31:31 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: connld.c,v $
 Revision 0.9.2.43  2008-09-22 20:31:31  brian
 - added module version and truncated logs

 Revision 0.9.2.42  2008-04-28 12:54:06  brian
 - update file headers for release

 *****************************************************************************/

#ident "@(#) $RCSfile: connld.c,v $ $Name:  $($Revision: 0.9.2.43 $) $Date: 2008-09-22 20:31:31 $"

static char const ident[] =
    "$RCSfile: connld.c,v $ $Name:  $($Revision: 0.9.2.43 $) $Date: 2008-09-22 20:31:31 $";

/* 
 *  This is CONNLD, a pipe module which generate new pipes for each open of an
 *  existing pipe-end.
 */

#include <linux/autoconf.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/kmem.h>
#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>

#include "sys/config.h"

#define CONNLD_DESCRIP		"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define CONNLD_COPYRIGHT	"Copyright (c) 2008-2009  Monavacon Limited.  All Rights Reserved."
#define CONNLD_REVISION		"LfS $RCSfile: connld.c,v $ $Name:  $($Revision: 0.9.2.43 $) $Date: 2008-09-22 20:31:31 $"
#define CONNLD_DEVICE		"SVR 4.2 CONNLD Module for STREAMS-based pipes"
#define CONNLD_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define CONNLD_LICENSE		"GPL"
#define CONNLD_BANNER		CONNLD_DESCRIP		"\n" \
				CONNLD_COPYRIGHT	"\n" \
				CONNLD_REVISION		"\n" \
				CONNLD_DEVICE		"\n" \
				CONNLD_CONTACT		"\n"
#define CONNLD_SPLASH		CONNLD_DEVICE		" - " \
				CONNLD_REVISION		"\n"

#ifdef CONFIG_STREAMS_CONNLD_MODULE
MODULE_AUTHOR(CONNLD_CONTACT);
MODULE_DESCRIPTION(CONNLD_DESCRIP);
MODULE_SUPPORTED_DEVICE(CONNLD_DEVICE);
MODULE_LICENSE(CONNLD_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-connld");
#endif
#ifdef MODULE_VERSION
MODULE_VERSION(__stringify(PACKAGE_RPMEPOCH) ":" PACKAGE_VERSION "." PACKAGE_RELEASE
	       PACKAGE_PATCHLEVEL "-" PACKAGE_RPMRELEASE PACKAGE_RPMEXTRA2);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_CONNLD_MODULE */

#ifndef CONFIG_STREAMS_CONNLD_NAME
//#define CONFIG_STREAMS_CONNLD_NAME "connld"
#error "CONFIG_STREAMS_CONNLD_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_CONNLD_MODID
//#define CONFIG_STREAMS_CONNLD_MODID 12
#error "CONFIG_STREAMS_CONNLD_MODID must be defined."
#endif

modID_t modid = CONFIG_STREAMS_CONNLD_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module ID for CONNLD.");

#ifdef MODULE_ALIAS
#ifdef LFS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_CONNLD_MODID));
MODULE_ALIAS("streams-module-connld");
#endif
#endif

static struct module_info connld_minfo = {
	.mi_idnum = CONFIG_STREAMS_CONNLD_MODID,
	.mi_idname = CONFIG_STREAMS_CONNLD_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat connld_mstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

/* 
 *  -------------------------------------------------------------------------
 *
 *  Open and close.
 *
 *  -------------------------------------------------------------------------
 */
static streamscall int
connld_open(queue_t *q, dev_t *devp, int oflag, int sflag, cred_t *credp)
{
	return (ENXIO);
}
static streamscall int
connld_close(queue_t *q, int oflag, cred_t *credp)
{
	return (ENXIO);
}
static streamscall int
connld_putp(queue_t *q, mblk_t *mp)
{
	putnext(q, mp);
	return (0);
}

/* 
 *  -------------------------------------------------------------------------
 *
 *  Registration and initialization
 *
 *  -------------------------------------------------------------------------
 */
static struct qinit connld_qinit = {
	.qi_putp = connld_putp,
	.qi_qopen = connld_open,
	.qi_qclose = connld_close,
	.qi_minfo = &connld_minfo,
	.qi_mstat = &connld_mstat,
};

static struct streamtab connld_info = {
	.st_rdinit = &connld_qinit,
	.st_wrinit = &connld_qinit,
};

static struct fmodsw connld_fmod = {
	.f_name = CONFIG_STREAMS_CONNLD_NAME,
	.f_str = &connld_info,
	.f_flag = D_MP,
	.f_kmod = THIS_MODULE,
};

#ifdef CONFIG_STREAMS_CONNLD_MODULE
static
#endif
int __init
connld_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_CONNLD_MODULE
	printk(KERN_INFO CONNLD_BANNER);
#else
	printk(KERN_INFO CONNLD_SPLASH);
#endif
	connld_minfo.mi_idnum = modid;
	if ((err = register_strmod(&connld_fmod)) < 0)
		return (err);
	if (modid == 0 && err > 0)
		modid = err;
	return (0);
};

#ifdef CONFIG_STREAMS_CONNLD_MODULE
static
#endif
void __exit
connld_exit(void)
{
	int err;

	if ((err = unregister_strmod(&connld_fmod)) < 0)
		return (void) (err);
	return (void) (0);
};

#ifdef CONFIG_STREAMS_CONNLD_MODULE
module_init(connld_init);
module_exit(connld_exit);
#endif
